### Dataset

>Dataset: [Turkish_Movie_Sentiment](https://www.win.tue.nl/~mpechen/projects/smm/)

>It consists of 5331 positive and 5331 negative movie reviews in Turkish.

### Load Dataset

> Load the dataset, unzip it, construct the dataframe out of it with respect to columns **[label, review]**.

> **You can use the curl command the download dataset from its source *(1)*. You can upload the dataset yourself *(2)*.**

In [1]:
# 1
!curl https://www.win.tue.nl/~mpechen/projects/smm/Turkish_Movie_Sentiment.zip -o ./Turkish_Movie_Sentiment.zip

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  520k  100  520k    0     0   461k      0  0:00:01  0:00:01 --:--:--  461k


In [2]:
# 2
#from google.colab import files
#uploaded = files.upload()

In [3]:
!unzip /content/Turkish_Movie_Sentiment.zip -d /content

Archive:  /content/Turkish_Movie_Sentiment.zip
  inflating: /content/tr_polarity.neg  
  inflating: /content/tr_polarity.pos  


> **Each line is a review.**

In [4]:
# Positive Reviews
with open("./tr_polarity.pos", "rb") as f:
  review_pos_list = f.read().decode("iso-8859-9").replace("\r", "").split("\n")

# Negative Reviews
with open("./tr_polarity.neg", "rb") as f:
  review_neg_list = f.read().decode("iso-8859-9").replace("\r", "").split("\n")

In [5]:
print(f"First 5 positive reviews of total {len(review_pos_list)}:\n ", *review_pos_list[:5], sep="\n")
print(f"\n\nFirst 5 negative review of total {len(review_neg_list)}:\n ", *review_neg_list[:5], sep="\n")

First 5 positive reviews of total 5332:
 
gerçekten harika bir yapim birçok kez izledim gene izlerim özgürlük askini ve ingilizlerin ne kadar vahset olduklarini gözler önüne seren bir film ve tabi ki ask.... 
her izledigimde hayranlik duydugum gerçek klasik diyebilecegimiz filmlerden . içinde teknik hatalar barindirsa bile sinema olgusunun en üst noktalarindan.. 
gerçekten tarihi savas filmleri arasinda tartismasiz en iyisi , 12 yil boyunca acaba ikincisi çekirimi diye bekledigim bir film ,belki william wallace babasinin ölümünden sonra amcasi yanina almisti onu yetistirmisti belki bunu anlatan mükkemmel bir filim olablilr=). 
aldigi ödülleri sonuna dek hak eden muhtesem bir basyapit . 
özgürlük denilince aklima gelen ilk film.bir basyapit.. 


First 5 negative review of total 5332:
 
giseye oynayan bir film.mel gibson'in oyunculugu yine çok kötü.film bastan sona duygu sömürüsü ama anlayan nerde!. 
bircok yonden sahip olduklari zayifliklari populerligi iyi kullanmasiyla gidermis zayif 

In [6]:
import pandas as pd

In [7]:
# Convert reviews to pandas dataframe
dataset = pd.DataFrame([[1, prev] for prev in review_pos_list if prev] + [[0, nrev] for nrev in review_neg_list if nrev] , columns=["label", "review"])
dataset

label                                             review
0          1  gerçekten harika bir yapim birçok kez izledim ...
1          1  her izledigimde hayranlik duydugum gerçek klas...
2          1  gerçekten tarihi savas filmleri arasinda tarti...
3          1  aldigi ödülleri sonuna dek hak eden muhtesem b...
4          1  özgürlük denilince aklima gelen ilk film.bir b...
...      ...                                                ...
10656      0  yarisina bile gelmeden sikilip biraktim,murat ...
10657      0  rezalet bir senaryo rezalet oyunculuklar(tuba ...
10658      0  nerden bulmuslar böyle yönetmeni oyuncuyu bast...
10659      0  konu:bilindik senaryo:basit kurgu:çakma geriye...
10660      0  agda sahnesini iyi pazarlayan , oyunculuk dene...

[10661 rows x 2 columns]

### Tokenization, Punctuations, Stop Words, Case Folding

> Tokeinize, remove punctuations and stopwords, apply case folding.

In [8]:
# nltk library for tokenizing, punctuations and stopwords
import nltk
nltk.download("punkt")

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [9]:
from nltk.corpus import stopwords;
nltk.download("stopwords");
stop_words = stopwords.words("turkish");

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [10]:
print(len(stop_words), stop_words)

53 ['acaba', 'ama', 'aslında', 'az', 'bazı', 'belki', 'biri', 'birkaç', 'birşey', 'biz', 'bu', 'çok', 'çünkü', 'da', 'daha', 'de', 'defa', 'diye', 'eğer', 'en', 'gibi', 'hem', 'hep', 'hepsi', 'her', 'hiç', 'için', 'ile', 'ise', 'kez', 'ki', 'kim', 'mı', 'mu', 'mü', 'nasıl', 'ne', 'neden', 'nerde', 'nerede', 'nereye', 'niçin', 'niye', 'o', 'sanki', 'şey', 'siz', 'şu', 'tüm', 've', 'veya', 'ya', 'yani']


In [11]:
from nltk.tokenize import word_tokenize
import re
# Tokenizing, removing punctuations and stop words, applying case folding
for idx, row in dataset.iterrows():
    dataset.at[idx, "review"] = [token.lower() for token in word_tokenize(re.sub(r"\.", " . ", row["review"]))  if token.isalpha() and (token.lower() not in stop_words)]

### Train-Test Splits

> Split dataset into 2 parts **train dataset** (80%) and **test dataset** (20%).

In [12]:
# Train Dataset
train_dataset = dataset.sample(frac=0.8, random_state=1)

In [13]:
train_dataset

label                                             review
1777       1  [iyi, film, noluyor, lan, demekten, filme, kon...
6818       0  [das, experimenti, yeniden, çekmisler, gerek, ...
1305       1  [yaw, herhangi, bir, kanal, filmi, yayinlasa, ...
6106       0  [filmi, dün, gece, izledim, bekledigim, çikmad...
1185       1   [tom, hanks, senaryo, iyi, film, zaten, asmisss]
...      ...                                                ...
7062       0  [acayip, zorlama, bir, film, olmus, puani, pop...
4690       1  [dvd, sini, nerdeyse, hafta, önce, izledim, sa...
6497       0  [serinin, iyi, filmiydi, serinin, ilk, filmi, ...
10558      0  [seyrettim, foktan, filimdi, gidip, korsanini,...
8334       0  [umdugum, bulmadim, kesinlikle, çocuklara, yön...

[8529 rows x 2 columns]

In [14]:
len(train_dataset)

8529

In [15]:
train_dataset["label"].value_counts()

label
1    4275
0    4254
Name: count, dtype: int64

In [16]:
# Test Dataset
test_dataset = dataset.drop(train_dataset.index)

In [17]:
test_dataset

label                                             review
0          1  [gerçekten, harika, bir, yapim, birçok, izledi...
2          1  [gerçekten, tarihi, savas, filmleri, arasinda,...
15         1  [önce, bir, çikarip, izledim, senedir, arada, ...
18         1  [iskoçyanin, evlatlari, baslayip, özgürlügü, a...
20         1  [kesinlikle, kaçirilmamasi, gerekne, filmlerin...
...      ...                                                ...
10639      0  [hafta, sonu, izledim, basrol, oyuncusunun, ka...
10644      0        [yakinda, apoya, film, çekilirse, sasirmam]
10647      0  [kendileri, inanacakki, karsiyida, inandirarak...
10650      0  [oldu, olacak, oscara, aday, gösterelim, filmm...
10656      0  [yarisina, bile, gelmeden, sikilip, biraktim, ...

[2132 rows x 2 columns]

In [18]:
len(test_dataset)

2132

In [19]:
test_dataset["label"].value_counts()

label
0    1076
1    1056
Name: count, dtype: int64

### Build Vocabulary

> Create a token set which holds all the tokens in train dataset and a token named as ***\<UNKNOWN>***, for tokens may appear in test dataset but does not appear in train dataset.

In [20]:
total_tokens = 0
token_set = set()
for _, tokens in train_dataset["review"].items():
  total_tokens += len(tokens)
  for token in tokens:
    token_set.add(token)

# Add <UNKNOWN> to token set for occurence of words that are not in corpus.
token_set.add("<UNKNOWN>")

In [21]:
print(f"Corpus size: {total_tokens}, Vocabulary size: {len(token_set)}\n")
print("First 10 tokens:")
i = 0
for token in token_set:
  if i >= 10:
    break
  print(token)
  i += 1

Corpus size: 145979, Vocabulary size: 25227

First 10 tokens:
kalkismaz
diyene
hissetmemistim
dokunmus
edilerek
amel
maya
aglariz
oyunculukla
ifade


### Vectorization


> token2idx: {token: token_idx}

> An array for keeping document frequency of every token in the token set. *`token_DFs[token_index]`*

> A matrix for holding term frequency of every token in the each train set review. *`token_by_review[token_index, document, index]`*

> Create a dataframe out of train dataset with respect to columns "label" and "vector". Vectors are created by calculating tf-idf for every token and have same dimension as size of token set.

In [22]:
# token-to-index mapping
token2idx = {}
for idx, term in enumerate(token_set):
  token2idx.update({term:idx})

In [23]:
# Length of each review token wise
train_dataset["length"] = train_dataset["review"].apply(lambda r: len(r))

In [24]:
train_dataset

label                                             review  length
1777       1  [iyi, film, noluyor, lan, demekten, filme, kon...       8
6818       0  [das, experimenti, yeniden, çekmisler, gerek, ...      14
1305       1  [yaw, herhangi, bir, kanal, filmi, yayinlasa, ...      27
6106       0  [filmi, dün, gece, izledim, bekledigim, çikmad...       9
1185       1   [tom, hanks, senaryo, iyi, film, zaten, asmisss]       7
...      ...                                                ...     ...
7062       0  [acayip, zorlama, bir, film, olmus, puani, pop...      10
4690       1  [dvd, sini, nerdeyse, hafta, önce, izledim, sa...      22
6497       0  [serinin, iyi, filmiydi, serinin, ilk, filmi, ...      33
10558      0  [seyrettim, foktan, filimdi, gidip, korsanini,...      31
8334       0  [umdugum, bulmadim, kesinlikle, çocuklara, yön...       5

[8529 rows x 3 columns]

In [25]:
import numpy as np

In [26]:
vocab_size = len(token_set)
number_of_reviews = train_dataset.shape[0]

# number of reviews that tokens appears in for each token
token_DFs = [0] * vocab_size

# matrix of number of occurences for each token in each review
# token_by_review[token_index, document_index]
token_by_review = np.zeros((vocab_size, number_of_reviews))

for idx, (_, row) in enumerate(train_dataset.iterrows()):

  # DF
  review_token_set = set(row["review"])
  for token in review_token_set:
    token_idx = token2idx[token]
    token_DFs[token_idx] += 1

  # TF
  for token in row["review"]:
    token_idx = token2idx[token]
    token_by_review[token_idx, idx] += 1

In [27]:
# Setting DF of <UNKNOWN> to smallest float possible so its idf can be very large
# which will make <UNKNOWN> significant
import sys
token_DFs[token2idx["<UNKNOWN>"]] = sys.float_info.min

In [28]:
print(token_DFs[:50])
print(token_by_review)

[1, 2, 1, 1, 1, 1, 1, 1, 9, 13, 1, 1, 1, 1, 16, 1, 1, 2, 1, 1, 8, 2, 1, 9, 33, 1, 5, 1, 4, 1, 4, 5, 1, 1, 1, 1, 4, 2, 1, 3, 1, 1, 1, 5, 1, 1, 1, 1, 1, 1]
[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]


In [29]:
import math

def get_tfidf(token_idx, review_idx, token_count):
  tf = token_by_review[token_idx, review_idx] / token_count
  idf = math.log(number_of_reviews / token_DFs[token_idx])
  return tf * idf

In [30]:
review_vectors = []
for idx, (_, row) in enumerate(train_dataset.iterrows()):
  review_vector = [0] * vocab_size
  review_tokens = set(row["review"])
  for token in review_tokens:
    token_idx = token2idx[token]
    review_vector[token_idx] += get_tfidf(token_idx, idx, row["length"])
  review_vectors.append([row["label"], review_vector])

train_vector_df = pd.DataFrame(review_vectors, columns=["label", "vector"])

In [31]:
train_vector_df

label                                             vector
0         1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
1         0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
2         1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
3         0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
4         1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
...     ...                                                ...
8524      0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
8525      1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
8526      0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0.1965538800865931...
8527      0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
8528      0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...

[8529 rows x 2 columns]

### Model Training

> Calculating a vector for positive reviews and a vector for negative reviews using element-wise addition.

In [32]:
import numpy as np

# Vectors for positive & negative sentiments
V = len(token_set)
pos_vector = np.array([0.0] * V) # Initial positive sentiment vector
neg_vector = np.array([0.0] * V) # Initial negative sentiment vector

for _, row in train_vector_df.iterrows():
  vector = np.array(row["vector"])
  if row["label"] == 0:
      neg_vector += vector
  else:
      pos_vector += vector

In [33]:
print(f"positive sentiment vector of size: {len(pos_vector)}\n", pos_vector[:100])
print(f"negative sentiment vector of size: {len(neg_vector)}\n", neg_vector[:100])

positive sentiment vector of size: 25227
 [ 0.53242514  0.8756084   0.50284597  0.64651624  0.53242514  0.
  0.          0.          2.55610442  1.93219753  0.          0.64651624
  0.56570171  0.29197508  6.16455575  0.39353163  0.53242514  0.
  0.          0.27427962  0.          0.25327516  0.          0.
  8.18423082  0.30170758  0.          0.          0.          0.3620491
  0.2322707   1.01182225  0.29197508  0.26621257  0.          0.
  1.04024091  0.          0.          2.02291171  0.26621257  0.
  0.          0.29767158  0.27427962  0.28285086  0.          0.64651624
  0.          0.          0.33523064  0.          0.          0.
  0.41343275  0.90512274  0.37713448  3.81688471  0.56570171  0.
  0.          0.39800382  0.72296501  0.83664349  0.32325812  0.
  0.          0.          0.          0.64651624  0.91438897  0.33523064
  0.24462777  0.45256137  0.          1.88466515  0.92867558  0.
  0.69650668  0.         14.5361764   0.          1.49251432  0.50284597
  0.77395

### Similarity Measure: Cosine

> Defining a function which calculates cosine similarity between vectors of the same dimension.

In [34]:
def cosine_sim(vector1, vector2):
  dot_product = np.dot(vector1, vector2)  # dot product of the vectors
  magnitude1 = np.linalg.norm(vector1)    # length of vector1
  magnitude2 = np.linalg.norm(vector2)    # length of vector2

  cosine_similarity = dot_product / (magnitude1 * magnitude2)
  return cosine_similarity

###Testing and Result

#### Calculate test vectors using only term frequencies.

In [35]:
# Document vectors of Test Set
test_vector_df = None
vocab_size = len(token_set)
test_doc_vectors = []

for _, row in test_dataset.iterrows():
  vector = [0] * vocab_size # initial vector
  for token in row["review"]:
    idx = token2idx.get(token, token2idx.get("<UNKNOWN>"))
    vector[idx] += 1

  # Normalize the vectors by dividing TFs to number of tokens in the review
  for token in set(row["review"]):
    idx = token2idx.get(token, token2idx.get("<UNKNOWN>"))
    vector[idx] /= len(row["review"])

  test_doc_vectors.append([row["label"], vector])

test_vector_df = pd.DataFrame(test_doc_vectors, columns= ["label", "vector"])

In [36]:
test_vector_df

label                                             vector
0         1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
1         1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
2         1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
3         1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
4         1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
...     ...                                                ...
2127      0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
2128      0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
2129      0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
2130      0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
2131      0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...

[2132 rows x 2 columns]

In [37]:
test_labels, test_vectors  = test_vector_df.iloc[:,0], test_vector_df.iloc[:,1]
test_labels.value_counts()

label
0    1076
1    1056
Name: count, dtype: int64

In [38]:
# make all predictions
predicted_labels = []
for vector in test_vectors:
  pos_sim = cosine_sim(np.array(vector), pos_vector)
  neg_sim = cosine_sim(np.array(vector), neg_vector)
  predicted_labels.append((1 if pos_sim > neg_sim else 0))

<ipython-input-34-c806a6ff87e3>:6: RuntimeWarning: invalid value encountered in scalar divide
  cosine_similarity = dot_product / (magnitude1 * magnitude2)


In [39]:
# number of predictions
len(predicted_labels)

2132

In [40]:
# check if prediction made correctly or not for each prediction
total_predictions = len(predicted_labels)
correct_predictions = 0
pos_pos = 0 # predicted pos, it is pos   (correct prediction)
neg_neg = 0 # predicted neg, it is neg   (correct prediction)
pos_neg = 0 # predicted pos, it is neg   (uncorrect prediction)
neg_pos = 0 # predicted neg, it is pos   (uncorrect prediction)
for idx ,prediction in enumerate(predicted_labels):
  label = test_labels[idx]

  if prediction and label:
    pos_pos += 1
  elif prediction and not label:
    pos_neg += 1
  elif not prediction and not label:
    neg_neg += 1
  else:
    neg_pos += 1

  correct_predictions += 1 if prediction == test_labels[idx] else 0

print(f"Success rate: {correct_predictions / total_predictions:.4g}")
print(f"\nPrediction\tLabel\n")
print(f"positive\tpositive\t{pos_pos / total_predictions:.4g}")
print(f"negative\tnegative\t{neg_neg / total_predictions:.4g}")
print(f"positive\tnegative\t{pos_neg / total_predictions:.4g}")
print(f"negative\tpositive\t{neg_pos / total_predictions:.4g}")

Success rate: 0.8316

Prediction	Label

positive	positive	0.4461
negative	negative	0.3856
positive	negative	0.1191
negative	positive	0.04925


#### Calculate test vectors using TF-IDF

> DF values are pulled from train dataset's DF array *(token_DFs)*.

In [41]:
test_TFs = [[0] * vocab_size for _ in range(test_dataset.shape[0])]
for idx, (_, row) in enumerate(test_dataset.iterrows()):
  for token in row["review"]:
    token_idx = token2idx.get(token, token2idx.get("<UNKNOWN>"))
    test_TFs[idx][token_idx] += 1

In [42]:
def get_tfidf_test(review_idx, token_idx, token_count):
  tf = test_TFs[review_idx][token_idx] / token_count
  idf = math.log(number_of_reviews / token_DFs[token_idx])
  return tf * idf

In [43]:
vocab_size = len(token_set)
test_doc_vectors = []
for idx, (_, row) in enumerate(test_dataset.iterrows()):
  vector = [0] * vocab_size
  for token in row["review"]:
    token_idx = token2idx.get(token, token2idx.get("<UNKNOWN>"))
    vector[token_idx] += get_tfidf_test(idx, token_idx, len(row["review"]))
  test_doc_vectors.append([row["label"], vector])

test_vector_df = pd.DataFrame(test_doc_vectors, columns= ["label", "vector"])

In [44]:
test_vector_df

label                                             vector
0         1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
1         1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
2         1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
3         1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
4         1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
...     ...                                                ...
2127      0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
2128      0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
2129      0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
2130      0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...
2131      0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...

[2132 rows x 2 columns]

In [45]:
test_labels, test_vectors  = test_vector_df.iloc[:,0], test_vector_df.iloc[:,1]
test_labels.value_counts()

label
0    1076
1    1056
Name: count, dtype: int64

In [46]:
# make all predictions
predicted_labels = []
for vector in test_vectors:
  pos_sim = cosine_sim(np.array(vector), pos_vector)
  neg_sim = cosine_sim(np.array(vector), neg_vector)
  predicted_labels.append((1 if pos_sim > neg_sim else 0))

<ipython-input-34-c806a6ff87e3>:6: RuntimeWarning: invalid value encountered in scalar divide
  cosine_similarity = dot_product / (magnitude1 * magnitude2)


In [47]:
# number of predictions
len(predicted_labels)

2132

In [48]:
# check if prediction made correctly or not for each prediction
total_predictions = len(predicted_labels)
correct_predictions = 0
pos_pos = 0 # predicted pos, it is pos   (correct prediction)
neg_neg = 0 # predicted neg, it is neg   (correct prediction)
pos_neg = 0 # predicted pos, it is neg   (uncorrect prediction)
neg_pos = 0 # predicted neg, it is pos   (uncorrect prediction)
for idx ,prediction in enumerate(predicted_labels):
  label = test_labels[idx]

  if prediction and label:
    pos_pos += 1
  elif prediction and not label:
    pos_neg += 1
  elif not prediction and not label:
    neg_neg += 1
  else:
    neg_pos += 1

  correct_predictions += 1 if prediction == test_labels[idx] else 0

print(f"Success rate: {correct_predictions / total_predictions:.4g}")
print(f"\nPrediction\tLabel\n")
print(f"positive\tpositive\t{pos_pos / total_predictions:.4g}")
print(f"negative\tnegative\t{neg_neg / total_predictions:.4g}")
print(f"positive\tnegative\t{pos_neg / total_predictions:.4g}")
print(f"negative\tpositive\t{neg_pos / total_predictions:.4g}")

Success rate: 0.6168

Prediction	Label

positive	positive	0.1318
negative	negative	0.485
positive	negative	0.0197
negative	positive	0.3635


### Conclusion
TF-IDF is used to calculate train dataset vectors. For this dataset calculating test vectors using only term frequencies (sr: 0.8316) appears to be more accurate than using tf-idf (sr: 0.6168).

sr: success rate
